test/apps/iodemo/io_demo.cc

```diff
@@ -832,132 +889,192 @@ class DemoClient : public P2pDemoCommon {
         return tv.tv_sec + (tv.tv_usec * 1e-6);
     }
 
-    bool run() {
-        std::vector<UcxConnection*> conn;
-        conn.resize(opts().servers.size());
-        for (size_t i = 0; i < conn.size(); i++) {
-            conn[i] = connect(opts().servers[i]);
-            if (!conn[i]) {
-                LOG << "Connect to server ["
-                    << opts().servers[i]
-                    << "] Failed!";
-                for (size_t j = 0; j < i; j++) {
-                    delete conn[j];
-                }
-                return false;
+    const std::string server_name(size_t server_index) {
+        std::stringstream ss;
+        ss << "server [" << server_index << "] " << opts().servers[server_index];
+        return ss.str();
+    }
+
+    void connect_failed(size_t server_index) {
+        server_info_t& server_info = _server_info[server_index];
+
+        ++server_info.retry_count;
+
+        UcxLog log(LOG_PREFIX);
+        log << "Connect to " << server_name(server_index) << " failed"
+            << " (retry " << server_info.retry_count;
+        if (opts().client_retries < std::numeric_limits<long>::max()) {
+            log << "/" << opts().client_retries;
+        }
+        log << ")";
+
+        if (server_info.retry_count >= opts().client_retries) {
+            /* If at least one server exceeded its retries, bail */
+            _status = CONN_RETRIES_EXCEEDED;
+        }
+    }
+
+    void connect_all(bool force) {
+        if (_active_servers.size() == _server_info.size()) {
+            // All servers are connected
+            return;
+        }
+
+        if (!force && !_active_servers.empty()) {
+            // The active list is not empty, and we don't have to check the
+            // connect retry timeout
+            return;
+        }
+
+        double curr_time = get_time();
+        if (curr_time < (_prev_connect_time + opts().client_timeout)) {
+            // Not enough time elapsed since previous connection attempt
+            return;
+        }
+
+        for (size_t server_index = 0; server_index < _server_info.size();
+             ++server_index) {
+            server_info_t& server_info = _server_info[server_index];
+            if ((server_info.conn != NULL) ||
```

Comment:
minor: or C++ like:
```
std::vector<server_info_t>::iterator it;
for (it = std::find_if(_server_info.begin(), _server_info.end(),
                       server_is_not_connected);
     it != _server_info.end();
     it = std::find_if(it, _server_info.end(), server_is_not_connected)) {
...
}
```

Reply:
IMO it will be more complicated / more code